databaker/loaders/xls.py

```
@@ -0,0 +1,221 @@
+
+        try:
+            self.workbook = get_workbook()
+        except NotImplementedError as e:
+            if not with_formatting_info:
+                raise
+            else:
+                with_formatting_info=False
+                self.workbook = get_workbook()
+
+        
+    def make_tables(self):
+        """ Return the sheets in the workbook. """
+        tables = [XLSRowSet(name, self.workbook.sheet_by_name(name), self.window)
+                for name in self.workbook.sheet_names()]
+        return tables
+
+
+class XLSRowSet(RowSet):
+    """ Excel support for a single sheet in the excel workbook. Unlike
+    the CSV row set this is not a streaming operation. """
+
+    def __init__(self, name, sheet, window=None):
+        self.name = name
+        self.sheet = sheet
+        self.window = window or 1000
+        super(XLSRowSet, self).__init__(typed=True)
+
+    def raw(self, sample=False):
+        """ Iterate over all rows in this sheet. Types are automatically
+        converted according to the excel data types specified, including
+        conversion of excel dates, which are notoriously buggy. """
+        num_rows = self.sheet.nrows
+        for rownum in range(min(self.window, num_rows) if sample else num_rows):
+            row = []
+            for colnum, cell in enumerate(self.sheet.row(rownum)):
+                try:
+                    row.append(XLSCell.from_xlrdcell(cell, self.sheet, colnum, rownum))
+                except InvalidDateError:
+                    raise ValueError("Invalid date at '%s':%d,%d" % (
+                        self.sheet.name, colnum+1, rownum+1))
+            yield row
+
+class XLSCell(Cell):
+    @staticmethod
+    def from_xlrdcell(xlrd_cell, sheet, col, row):
+        value = xlrd_cell.value
+        cell_type = XLS_TYPES.get(xlrd_cell.ctype, StringType())
+    
+        if cell_type == DateType(None):
+            if value == 0:
+                raise InvalidDateError
```

if you replace the definition `InvalidDateError` at the top of this file (there's another comment on that, probably needs to be moved) with the following

```
class InvalidDateError(Exception):
    """ Raised when a cell_type has been identified as DateType but contains an invalid value
    """

    def __init__(self, message):
        self.message = message
```

then you can raise it with a message string to provide some context, eg `InvalidDateError(f'the date "{str()value}" is not a valid date.')` which'll make debugging weirdness a lot easier.